### PR TITLE
Refactor project structure and introduce composite builds

### PR DESCRIPTION
**Pull request by IgnatBeresnev**
At the moment, the project's build configuration is very coupled, which becomes problematic if one of the subprojects (like the Gradle plugin) needs to be more independent and be configured differently than the rest of the subprojects.

This PR serves as preparation work for #3131 (it will help with merging [Dokkatoo](https://github.com/adamko-dev/dokkatoo) into Dokka), and at the same time addresses several other issues:
* Make use of Gradle's [Composite builds](https://docs.gradle.org/current/userguide/composite_builds.html), which are configured and executed in isolation.
* Refactor the project structure:
    * Re-group the subprojects based on responsibility / layer [(analysis, core, plugin, runner).
    * Flatten subprojects (i.e `core/test-api` should become `core-test-api`)
* Some refactoring of the build scripts to be more idiomatic, which will help with adopting modern Gradle features

Not in scope:
* Renaming the artifacts (will be addressed later in #3245)
* Doing any sorts of improvements that would break backward compatibility for Dokka's users

___

This PR will be in the Draft state until the build is green and the tests pass. Should try to keep the commits more isolated so that they can be reviewed separately.

### CODE REVIEW

**Comment by IgnatBeresnev** [comment]
I've relocated all of the currently existing subprojects to adhere to the desired structure, so now we just gotta make it work :sweat_smile:

**Comment by IgnatBeresnev** [comment]
The PR is ready for review and ready to be merged.

Unfortunately, because everything is so tightly coupled and outdated, the scope grew and this PR also includes:
* Publishing refactoring, as described in #3132.
* Update to Gradle 8.4, along with the required fixes

## How to review it

### Diff

It may seem like a lot of files have been changed, but the overwhelming majority of them were simply moved with `git mv`, which preserves history. If you exclude the commits that move files and resolve merge conflicts, you will end up with a reviewable list of changes. 

I've excluded meaningless commits and generated a `.patch` file, and thankfully GitHub's gists know how to render them in a nice way, so you can have a look at the following gist for meaningful changes in this PR: https://gist.github.com/IgnatBeresnev/3ed43625bc01ea22421290458b9b2755

Another way to review it is to open it in IDEA and only select the following commits: 

<details>

<summary>List of meaningful commits</summary>

* Update hardcoded project paths
* Update coroutine's diff
* Fix Dokka version in coroutines and serialization tests
* Fix integration tests
* Fix publishing after the update to Gradle 8
* Add parent validatePlugins task
* Fix CLI integration tests
* Make dokka-subproject a non-included build
* Return a lost `@Optional` annotation
* Make integration tests runnable from IDEA
* Enable parent publishPlugins task
* Remove `gradle-plugin` subproject to minimize diff
* Minor build scripts refactoring
* Minor convention plugin refactoring
* Rename convention plugins
* Remove check's dependency on integration tests
* Replace deprecated API
* fix s3 script path
* run clean task separately
* ignore some tests for now
* fix gradle plugin compatibility
* restore BCV
* fix workflows
* update gradle to 8.4 (not fully yet)
* add parent tasks to dokka-integration-tests
* cli integration tests
* fix maven plugin POM
* maven integration tests
* rollback gradle integration tests to use mavenLocal for now
* restore gradle integration tests
* remove todos
* fix publishing of Gradle plugin and it markers
* cleanup build scripts setup group and version in properties setup publishing for gradle plugin
* Correct publishing information
* Add parent composite build tasks
* Correct publishing env properties
* Setup publishing to multiple repositories (gradle-plugin publishing is not supported yet)
* Rip out publishing
* Rip out non essentials
* Fix `dokka-core` artifact name for runners
* Move gradle-plugin's versions into the common catalog
* Rip out non-essential dependencies in gradle-plugin
* Align gradle-plugin's formatting with the rest of the project
* Move parts of Dokkatoo into gradle-plugin
* Align included build toolchain and language versions
* Add composite build parent tasks
* Fix Gradle sync problems
* Update dokka_version to 1.9.20-SNAPSHOT
* Remove irrelevant project properties
* Add Gradle properties to included builds
* Add version-catalog def in integration-tests
* Initial included-build refactor

</details>

## Places in code

If you'd prefer not to look at the diff and look at the code instead, the only really meaningful directories and files that need to be reviewed:

* The overall project structure (where the modules are and how they are named)
* All convention plugins inside `~/IdeaProjects/dokka/build-logic/src/main/kotlin`
* `~/IdeaProjects/dokka/build.gradle.kts`

The rest of the changes were necessary to make the project buildable and pass integration tests.

File: build-logic/src/main/kotlin/dokkabuild.publish-gradle-plugin.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2014-2023 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+import dokkabuild.PublicationName
+
+/*
```
[the remark this answers is not in the document]
Comment (IgnatBeresnev) [reply build-logic/src/main/kotlin/dokkabuild.publish-gradle-plugin.gradle.kts:7]: Indeed, removed this one and checked all `.kts` files as well

File: dokka-integration-tests/cli/build.gradle.kts [hunk 2]
```diff
@@ -5,17 +5,14 @@
...
 }
 
-val dokka_version: String by project
-evaluationDependsOn(":runners:cli")
-evaluationDependsOn(":plugins:base")
-
 dependencies {
     implementation(kotlin("test-junit5"))
-    implementation(projects.integrationTests)
+    implementation(libs.junit.jupiterApi)
```
[the remark this answers is not in the document]
Comment (IgnatBeresnev) [reply dokka-integration-tests/cli/build.gradle.kts:14]: Good question. I was the one who added it, and I think it didn't work without it, but maybe with include builds or new gradle it's not needed anymore, dunno. I've removed it and everything seems to work :man_shrugging:

File: dokka-integration-tests/gradle/projects/template.settings.gradle.kts [hunk 3]
```diff
@@ -31,7 +33,16 @@ pluginManagement {
         }
     }
     repositories {
-        mavenLocal()
+        mavenLocal {
+            content {
+                includeGroup("org.jetbrains.dokka")
+            }
+        }
+        mavenLocal {
```
[the remark this answers is not in the document]
Comment (IgnatBeresnev) [reply dokka-integration-tests/gradle/projects/template.settings.gradle.kts:41]: Removed it

**Review by whyoleg — APPROVED** [review]
Let's finally merge it :)

**Pull request merged**